### PR TITLE
Enable Deepspeed Ulysses for Wan

### PR DESCRIPTION
**Pull request by mengker33**

### CODE REVIEW

File: DeepSpeed/deepspeed/sequence/layer.py [hunk 1]
```diff
@@ -184,7 +184,7 @@ def single_all_to_all(input, scatter_idx, gather_idx, batch_dim_idx, group, asyn
     # we only need num_heads once
     num_heads = input.shape[2]
 
-    if get_num_kv_heads() is not None or num_heads % seq_world_size != 0:
+    if get_num_kv_heads() is not None or (num_heads % seq_world_size != 0 and not scatter_idx < 2):
```
Comment (ikurtchen) [inline DeepSpeed/deepspeed/sequence/layer.py:187]: What's the case for adding this check? Will uneven heads use even all2all in this case?

File: examples/stable-diffusion/image_to_video_generation.py [hunk 2]
```diff
@@ -295,6 +304,14 @@ def main():
     if args.bf16:
         kwargs["torch_dtype"] = torch.bfloat16
 
+    if args.context_parallel_size > 1 and parallel_state.is_unitialized():
```
Comment (ikurtchen) [inline examples/stable-diffusion/image_to_video_generation.py:307]: Is this a typo? "is_unitialized", is_uninitialized?

File: optimum/habana/diffusers/models/attention_processor.py [hunk 3]
```diff
@@ -206,8 +208,94 @@ def __init__(self, fusedSDPA):
...
+            recompute_mode,
+            valid_sequence_lengths,
+            padding_side,
+        )
+        return out.permute(0, 2, 1, 3)
+
+
+class GaudiDistributedAttention(torch.nn.Module):
+    def __init__(
+        self, hpu_module_fsdpa: ModuleFusedSDPA, scale, attention_dropout, enable_recompute, flash_attention_fp8
```
Comment (ikurtchen) [inline optimum/habana/diffusers/models/attention_processor.py]: scale, attention_dropout, enable_recompute, flash_attention_fp8 are not used?

File: optimum/habana/diffusers/models/attention_processor.py [hunk 4]
```diff
@@ -565,14 +668,27 @@ def _native_attention(
...
-        from habana_frameworks.torch.hpex.kernels import FusedSDPA
-
         # Fast FSDPA is not supported in training mode
         fsdpa_mode = "None" if self.is_training else "fast"
-        query, key, value = (x.permute(0, 2, 1, 3) for x in (query, key, value))
-        out = FusedSDPA.apply(query, key, value, attn_mask, dropout_p, is_causal, scale, fsdpa_mode, None)
-        out = out.permute(0, 2, 1, 3)
+
+        if self.fused_scaled_dot_product_attention_distributed:
+            out = self.fused_scaled_dot_product_attention_distributed(
```
Comment (ikurtchen) [inline optimum/habana/diffusers/models/attention_processor.py:673]: Why the FusedSDPA call is different from not SP case?

File: optimum/habana/diffusers/pipelines/wan/pipeline_wan_i2v.py [hunk 5]
```diff
@@ -176,6 +177,22 @@ def prepare_latents(
         latent_height = height // self.vae_scale_factor_spatial
         latent_width = width // self.vae_scale_factor_spatial
 
+        def adjust_num_frame(t, h, w, cp_size):
+            # It is to make sure seq_len can be devided by cp_size.
+            MAX_T = t + 100
```
Comment (ikurtchen) [inline optimum/habana/diffusers/pipelines/wan/pipeline_wan_i2v.py]: Is it possible not hard code 100 to find the padding size?

File: optimum/habana/diffusers/pipelines/wan/pipeline_wan_i2v.py [hunk 6]
```diff
@@ -176,6 +177,22 @@ def prepare_latents(
...
+            for t_i in range(t, MAX_T):
+                if (t_i * h * w) % cp_size == 0:
+                    return t_i
+            raise ValueError(f"No valid num_latent_frames found up to {MAX_T} for h={h}, w={w}, cp_size={cp_size}")
+
+        cp_size = parallel_state.get_sequence_parallel_world_size()
+        p_t, p_h, p_w = self.transformer.config.patch_size
+        tmp_seq_len = (num_latent_frames // p_t) * (latent_height // p_h) * (latent_width // p_w)
+        if tmp_seq_len % cp_size != 0:
+            num_latent_frames = adjust_num_frame(
```
Comment (ikurtchen) [inline optimum/habana/diffusers/pipelines/wan/pipeline_wan_i2v.py]: Dose it need to remember how many padding added here, so as to mask attn and remove padding later?

File: DeepSpeed/deepspeed/sequence/layer.py (hunk 1 above)
Comment (mengker33) [reply DeepSpeed/deepspeed/sequence/layer.py:187]: For last output all2all, head is split and seq is complete, in this case there is no need to check num_heads % seq_world_size. See https://github.com/HabanaAI/optimum-habana-fork/blob/aice/v1.22.0/DeepSpeed/deepspeed/sequence/layer.py#L405.

File: examples/stable-diffusion/image_to_video_generation.py (hunk 2 above)
Comment (mengker33) [reply examples/stable-diffusion/image_to_video_generation.py:307]: It is a typo from oh itself and I don't prefer to correct it here. See https://github.com/HabanaAI/optimum-habana-fork/blob/aice/v1.22.0/optimum/habana/distributed/parallel_state.py#L353

File: optimum/habana/diffusers/models/attention_processor.py (hunk 3 above)
Comment (mengker33) [reply optimum/habana/diffusers/models/attention_processor.py]: You're right, let me remove these unused parameters.

File: optimum/habana/diffusers/models/wan_transformer_3d.py [hunk 7]
```diff
@@ -66,6 +68,25 @@ def WanTransformer3DModleForwardGaudi(
     hidden_states = self.patch_embedding(hidden_states)
     hidden_states = hidden_states.flatten(2).transpose(1, 2)
 
+    if parallel_state.sequence_parallel_is_initialized():
+        seq_len = hidden_states.shape[1]
+        sp_seq_len = seq_len // parallel_state.get_sequence_parallel_world_size()
+        start = sp_seq_len * parallel_state.get_sequence_parallel_rank()
+        end = sp_seq_len * (parallel_state.get_sequence_parallel_rank() + 1)
+
+        hidden_states = hidden_states[:, start:end, :]
```
Comment (Wei-Lin-Intel) [inline optimum/habana/diffusers/models/wan_transformer_3d.py:101]: Padded sequence length in SDPA without a mask would trigger the accuracy issue. It is necessary to add mask here if `seq_len` is padded. Use bool tensor for the mask can save some memory.

File: optimum/habana/diffusers/pipelines/wan/pipeline_wan_i2v.py (hunk 5 above)
Comment (mengker33) [reply optimum/habana/diffusers/pipelines/wan/pipeline_wan_i2v.py]: Move the padding after patching.

File: optimum/habana/diffusers/pipelines/wan/pipeline_wan_i2v.py (hunk 6 above)
Comment (mengker33) [reply optimum/habana/diffusers/pipelines/wan/pipeline_wan_i2v.py]: I have added the attention_mask after padding. But considering padding size is very small compared to original seq_len, mask will not be applied by default.

File: optimum/habana/diffusers/models/attention_processor.py [hunk 8]
```diff
@@ -206,8 +208,92 @@ def __init__(self, fusedSDPA):
...
+        attn_mask,
+        dropout_p,
+        is_causal,
+        scale,
+        softmax_mode,
+        recompute_mode,
+        valid_sequence_lengths,
+        padding_side="left",
+    ):
+        query, key, value = (x.permute(0, 2, 1, 3) for x in (query, key, value))
```
Comment (yingjie-han) [inline optimum/habana/diffusers/models/attention_processor.py]: 单卡跑的时候，这里比原始代码多了permute。fused_scaled_dot_product_attention输入的时候先transpose,这里在permute，这样是否会带来不必要的开销。

Comment (Wei-Lin-Intel) [reply optimum/habana/diffusers/models/attention_processor.py]: Add a `.contiguous()` after permute may avoid DMA cost in SDPA

File: optimum/habana/diffusers/models/attention_processor.py [hunk 9]
```diff
@@ -262,17 +348,20 @@ def __call__(
...
         hidden_states = self.fused_scaled_dot_product_attention(
-            query,
-            key,
-            value,
-            attn_mask=attention_mask,
-            dropout_p=0.0,
-            is_casual=False,
-            scale=None,
-            softmax_mode=softmax_mode,
+            query.transpose(1, 2),
```
Comment (Wei-Lin-Intel) [inline optimum/habana/diffusers/models/attention_processor.py]: same here, add `.contiguous()` here

File: optimum/habana/diffusers/models/attention_processor.py (hunk 8 above)
Comment (mengker33) [reply optimum/habana/diffusers/models/attention_processor.py]: It is moving permute from ModuleSDPA outside to inside, the same for single-card and multi-card.

File: optimum/habana/diffusers/models/attention_processor.py (hunk 4 above)
Comment (mengker33) [reply optimum/habana/diffusers/models/attention_processor.py:673]: make them the aligned call.

**Comment by huijuanzh** [comment]
LGTM, if update some performance data compared with 1/2/4 card would be good.

File: optimum/habana/diffusers/models/wan_transformer_3d.py [hunk 10]
```diff
@@ -66,6 +70,47 @@ def WanTransformer3DModleForwardGaudi(
...
+                attention_mask = torch.ones(bs, 1, seq_len, seq_len, dtype=hidden_states.dtype, device=hidden_states.device)
+                attention_mask = F.pad(attention_mask, (0, pad_len, 0, pad_len)).bool()
+
+            seq_len = padded_seq_len
+
+        sp_seq_len = seq_len // parallel_state.get_sequence_parallel_world_size()
+        start = sp_seq_len * parallel_state.get_sequence_parallel_rank()
+        end = sp_seq_len * (parallel_state.get_sequence_parallel_rank() + 1)
+
+        hidden_states = hidden_states[:, start:end, :]
```
Comment (yingjie-han) [inline optimum/habana/diffusers/models/wan_transformer_3d.py:101]: Here attention_mask is not splitted. I think attention_mask also need to be splited by sp , what do you think?

Comment (mengker33) [reply optimum/habana/diffusers/models/wan_transformer_3d.py:101]: attention_mask is only used in deepspeed ulysses self-attention case where in ModuleFusedSDPA, q/k/v are already gathered along seq and split along head, so it is ok to create the attention_mask  with full seq len here.

**Comment by mengker33** [comment]
> LGTM, if update some performance data compared with 1/2/4 card would be good.

Linwei suggested to use classic SP to reduce memory consumption. I will include SP in this PR too and provide perf data for both deepspeed ulysses and SP soon.

**Comment by mengker33** [comment]
image
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>


<div style='direction:ltr'>


num_card | SP:   generated_time | DS:   generated_time
-- | -- | --
1 | 245s | 245s
2 | Avg:   145s | Avg:141s
4 | Avg:74.75s | Avg: 71s
8 | Avg: 59s | Avg: 55s



</div>


</body>

</html>

@Wei-Lin-Intel @huijuanzh Please help to review again, thank you:)

**Review by Wei-Lin-Intel — APPROVED** [review]
LGTM

File: optimum/habana/diffusers/models/attention_processor.py [hunk 11]
```diff
@@ -262,17 +348,20 @@ def __call__(
...
         hidden_states = self.fused_scaled_dot_product_attention(
-            query,
-            key,
-            value,
-            attn_mask=attention_mask,
-            dropout_p=0.0,
-            is_casual=False,
-            scale=None,
-            softmax_mode=softmax_mode,
+            query.transpose(1, 2).contiguous(),
```
Comment (huijuanzh) [inline optimum/habana/diffusers/models/attention_processor.py:351]: why CogVideoXAttnProcessorGaudi need to change? Since you put qkv transpose in ModuleFusedSDPA, need to check other function which not use sp or ulysses.

**Comment by huijuanzh** [comment]
LGTM
